drivers/serial/uart_nrfx_uarte.c

```diff
@@ -39,39 +39,52 @@ LOG_MODULE_REGISTER(uart_nrfx_uarte, CONFIG_UART_LOG_LEVEL);
 #error "No PPI or DPPI"
 #endif
 
+/* We cannot use LISTIFY because LISTIFY cannot be used recursively and it is already
+ * used inside macros used by UARTE device instantiation so we list all existing
+ * IDs used for UARTE peripheral in all Nordic devices.
+ */
+#define ALL_UARTE_INSTANCES \
+	0, 1, 2, 3, 00, 20, 21, 22, 30, 120, 130, 131, 132, 133, 134, 135, 136, 137
+
+/* Execute macro f(x) for all instances. */
+#define UARTE_FOR_EACH_INSTANCE(f, sep, off_code) \
+	NRFX_FOREACH_PRESENT(UARTE, f, sep, off_code, _)
 
-#if	(defined(CONFIG_HAS_HW_NRF_UARTE0) &&         \
-	 defined(CONFIG_UART_0_INTERRUPT_DRIVEN)) || \
-	(defined(CONFIG_HAS_HW_NRF_UARTE1) &&         \
-	 defined(CONFIG_UART_1_INTERRUPT_DRIVEN)) || \
-	(defined(CONFIG_HAS_HW_NRF_UARTE2) &&         \
-	 defined(CONFIG_UART_2_INTERRUPT_DRIVEN)) || \
-	(defined(CONFIG_HAS_HW_NRF_UARTE3) &&         \
-	 defined(CONFIG_UART_3_INTERRUPT_DRIVEN))
+/* Determine if any instance is using interrupt driven API. */
+#define IS_INT_DRIVEN(unused, prefix, i, _) IS_ENABLED(CONFIG_UART_##prefix##i##_INTERRUPT_DRIVEN)
```

Comment:
I think you need to use `IS_ENABLED(CONFIG_HAS_HW_NRF_UARTE##prefix##i) &&` here because of that `CONFIG_UART_0_INTERRUPT_DRIVEN` option that can be also enabled for `CONFIG_HAS_HW_NRF_UART0`.